src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -1941,6 +1958,155 @@ void sort(json::Value& val) {
+        // know which keyword to use to get the base_uri
+        auto id_it = [&] {
+            switch (maybe_new_dialect.value()) {
+            case json_schema_dialect::draft4:
+                return maybe_draft4_id_it;
+            case json_schema_dialect::draft6:
+            case json_schema_dialect::draft7:
+            case json_schema_dialect::draft201909:
+            case json_schema_dialect::draft202012:
+                return maybe_id_it;
+            }
+        }();
+
+        if (id_it == this_obj.object_range().end()) {
+            // stop scanning this branch, the keyword for base uri does not
+            // agree with schema dialect.
+            return;
+        }
+
+        // run validation since we are not a guaranteed to be in proper schema
+        if (validate_json_schema(maybe_new_dialect.value(), this_obj)
+              .has_error()) {
+            // stop exploring this branch, the schema is invalid
+            return;
+        }
+
+        // base uri keyword agrees with the dialect, it's a validated schema, we
+        // can register this as a bundled schema and continue scanning.
+        // (run resolve because it could be relative to the parent schema).
+        base_uri = jsoncons::uri{id_it->value().as_string()}.resolve(base_uri);
+        dialect = maybe_new_dialect.value();
+        bundled_schemas.insert_or_assign(
+          to_json_id_uri(base_uri),
+          std::pair{json::Pointer{this_obj_ptr.to_string()}, dialect});
+    }
+
+    if (auto ref_it = this_obj.find("$ref");
+        ref_it != this_obj.object_range().end()) {
+        // ensure refs are absolute uris
+        ref_it->value() = jsoncons::uri{ref_it->value().as_string()}
+                            .resolve(base_uri)
+                            .string();
+    }
+
+    // lambda to recursively scan the object for more bundled schemas and $refs
+    auto collect_and_fix = [&](const auto& key, auto& value) {
+        collect_bundled_schemas_and_fix_refs(
+          bundled_schemas, base_uri, this_obj_ptr / key, value, dialect);
+    };
+
+    // recursively scan the object for more bundled schemas and $refs
+    for (auto& e : this_obj.object_range()) {
+        const auto& key = e.key();
+        auto& value = e.value();
+        if (value.is_object()) {
+            collect_and_fix(key, value);
+        } else if (value.is_array()) {
+            for (auto i = 0u; i < value.size(); ++i) {
+                if (value[i].is_object()) {
+                    collect_and_fix(i, value[i]);
+                }
+            }
+        }
+    }
+}
+
+id_to_schema_pointer collect_bundled_schema_and_fix_refs(
+  jsoncons::json& doc, json_schema_dialect dialect) {
+    // entry point to collect all bundled schemas
+    auto bundled_schemas = id_to_schema_pointer{};
+    if (doc.is_bool()) {
+        bundled_schemas.insert_or_assign(
+          {}, std::pair{json::Pointer{}, dialect});
```

Comment:
i wanted to keep the invariant of bundled schemas always having at least an entry for the root. 
in this specific case there is no practical purpose, but since in the rest of the file we try to treat `true` as `{}` i felt that it's no harm to extend this also here.